### PR TITLE
Better API key generation

### PR DESCRIPTION
**Pull request by alexander-yu**
Improved the method for generating API keys (instead of just using `os.random` and turning it into a string). Additionally, corresponding UNIQUE index + constraint for the `code` column (i.e. API key) the `oauth_data` table was added to reinforce the fact that API keys need to be unique.

Also updated build status badge in the README.

### CODE REVIEW

File: config/oauth_dev_dump.sql [hunk 1]
```diff
@@ -10,6 +10,9 @@ CREATE TABLE oauth_data (
   code text NOT NULL
 );
 
+CREATE UNIQUE INDEX on oauth_data (code);
+ALTER TABLE oauth_data ADD CONSTRAINT unique_code UNIQUE (code);
```
Comment (alanhdu) [inline config/oauth_dev_dump.sql]: Do we actually need both statements? Seems redundant to me.

File: config/schema.sql [hunk 2]
```diff
@@ -103,6 +103,9 @@ CREATE TABLE oauth_data (
     code text NOT NULL
 );
 
+CREATE UNIQUE INDEX on oauth_data (code);
+ALTER TABLE oauth_data ADD CONSTRAINT unique_code UNIQUE (code);
```
Comment (alanhdu) [inline config/schema.sql]: Same here.

File: density/db/db.py [hunk 3]
```diff
@@ -188,8 +189,8 @@ def get_oauth_code_for_uni(cursor, uni):
         return result['code']
     else:
         # If the code DNE, create a new one and insert into the database.
-        new_code = ''.join(random.choice(
-            string.ascii_uppercase + string.digits) for x in xrange(32))
+        token_bytes = os.urandom(32) + uuid.uuid4().bytes
+        new_code = base64.urlsafe_b64encode(token_bytes)
         query = """INSERT INTO oauth_data (uni, code)
                    VALUES (%s, %s);"""
         cursor.execute(query, [uni, new_code])
```
Comment (alanhdu) [inline density/db/db.py]: There should probably be some kind of exception handling in case the `new_code` is not unique. (For completeness, you know. Even though we'd probably never run into it).

File: density/db/db.py [hunk 4]
```diff
@@ -188,8 +189,8 @@ def get_oauth_code_for_uni(cursor, uni):
         return result['code']
     else:
         # If the code DNE, create a new one and insert into the database.
-        new_code = ''.join(random.choice(
-            string.ascii_uppercase + string.digits) for x in xrange(32))
+        token_bytes = os.urandom(32) + uuid.uuid4().bytes
+        new_code = base64.urlsafe_b64encode(token_bytes)
```
Comment (alanhdu) [inline density/db/db.py]: If this string has fixed-length (which I think it does?), then we should make the `code` column a `VARCHAR` instead of `TEXT`

File: config/schema.sql (hunk 2 above)
Comment (alexander-yu) [inline config/schema.sql]: It's not necessary to have both, but momentarily just added both lines since I wasn't sure what the preference between the two would be. Might be better just to keep the INDEX.

File: density/db/db.py (hunk 3 above)
Comment (alexander-yu) [inline density/db/db.py]: Yeah, I was thinking of having some error the small chance it does happen; maybe some message displayed that asks the user to refresh the page to generate a new key?

File: density/db/db.py (hunk 4 above)
Comment (alexander-yu) [inline density/db/db.py]: Yeah, it's a fixed-length key; sounds good to me.

File: config/schema.sql (hunk 2 above)
Comment (alanhdu) [inline config/schema.sql]: I think we should go with the `CREATE UNIQUE INDEX` -- it seems more obviously like what we want. It's also unclear to me that a `UNIQUE` constraint is the same thing as a `UNIQUE INDEX` (I'm pretty sure they are in MySQL, but I'm not sure about Postgres). After all, a `UNIQUE` constraint doesn't need to speed-up querying, which probably opens up the design space a little (I'm imagining a hash table w/ no range queries for the constraint vs a  b-tree for the index).

File: density/db/db.py (hunk 3 above)
Comment (alanhdu) [inline density/db/db.py]: Either that or a `while` loop. Doesn't really matter to me -- I just don't want the app to crash (I'm not sure how we handle restarting apps in production).

File: density/db/db.py (hunk 4 above)
Comment (alanhdu) [inline density/db/db.py]: Awesome. Honestly, I didn't even know you _could_ create an index on `TEXT`

File: config/schema.sql (hunk 2 above)
Comment (alexander-yu) [inline config/schema.sql]: I think from looking at the PostgresSQL documentation, it seems that a `UNIQUE CONSTRAINT` under the hood is the same as a `UNIQUE INDEX` (i.e. creates a B-tree index), but it has some additional checking for uniqueness, and it appears the two have some differences in terms of features/options.

Regardless, an `INDEX` is probably the better option since we're actually going to be using it as an index (also makes it clearer what the intended usage is for); I'll go ahead and get rid of the `CONSTRAINT` portion.

File: density/db/db.py (hunk 3 above)
Comment (alexander-yu) [inline density/db/db.py]: Yeah, fair enough; I'll get on that.

**Comment by alanhdu** [comment]
Do you mind squashing some commits? (For example, the two index commits should be combined, and the test fix should probably be squashed too).

Otherwise, LGTM

File: density/db/db.py [hunk 5]
```diff
@@ -188,8 +189,8 @@ def get_oauth_code_for_uni(cursor, uni):
...
         query = """INSERT INTO oauth_data (uni, code)
                    VALUES (%s, %s);"""
```
Comment (alanhdu) [inline density/db/db.py]: ... We really should audit all the functions here for sql injections, but let's leave that for another PR.

**Comment by alexander-yu** [comment]
No problem; commits squashed.

File: density/db/db.py (hunk 5 above)
Comment (alexander-yu) [inline density/db/db.py]: Yeah, that would definitely be a good idea; I'm not very familiar with injection-safe SQL practices (to be honest I'm not super familiar with SQL in general), but I can look into it at one point.
